src/main/java/org/dataloader/DataLoaderHelper.java

```diff
@@ -546,4 +617,197 @@ private CompletableFuture<List<V>> setToValueCache(List<V> assembledValues, List
     private static <T> DispatchResult<T> emptyDispatchResult() {
         return (DispatchResult<T>) EMPTY_DISPATCH_RESULT;
     }
+
+    private class DataLoaderSubscriber implements Subscriber<V> {
+
+        private final CompletableFuture<List<V>> valuesFuture;
+        private final List<K> keys;
+        private final List<Object> callContexts;
+        private final List<CompletableFuture<V>> queuedFutures;
+
+        private final List<K> clearCacheKeys = new ArrayList<>();
+        private final List<V> completedValues = new ArrayList<>();
+        private int idx = 0;
+        private boolean onErrorCalled = false;
+        private boolean onCompleteCalled = false;
+
+        private DataLoaderSubscriber(
+            CompletableFuture<List<V>> valuesFuture,
+            List<K> keys,
+            List<Object> callContexts,
+            List<CompletableFuture<V>> queuedFutures
+        ) {
+            this.valuesFuture = valuesFuture;
+            this.keys = keys;
+            this.callContexts = callContexts;
+            this.queuedFutures = queuedFutures;
+        }
+
+        @Override
+        public void onSubscribe(Subscription subscription) {
+            subscription.request(keys.size());
+        }
```

Comment:
This is the most sensible thing - to ask for all keys - I had the thought that setup might want to control how many get asked at a time BUT we need to get `key.size()` amount and they will all be held in memory anyway so getting them all is the most sensible thing to do